library/ssl_tls.c

```diff
@@ -1354,7 +1354,7 @@ void mbedtls_ssl_conf_session_cache( mbedtls_ssl_config *conf,
 int mbedtls_ssl_set_session( mbedtls_ssl_context *ssl, const mbedtls_ssl_session *session )
 {
     int ret = MBEDTLS_ERR_ERROR_CORRUPTION_DETECTED;
-
+    mbedtls_ssl_session *target = ssl->session_negotiate;
```

Comment:
Eventually I think that copying the first session resumption candidate (session passed in the first call to `mbedtls_ssl_set_session`) in `ssl->session_negotiate` makes things more complicated and error prone.

In the multiple session case, that is not this particular session that we want to resume but one of the sessions in a list of sessions. Thus eventually I think I would be rather on the side of adding a new `mbedtls_ssl_session *resumption_session_list` field in the `mbedtls_ssl_handshake_params` structure and store the address of the head of the resumption session list here.

Then the `mbedtls_ssl_session_free_chain` (probably better to rename it `mbedtls_ssl_session_list_free`) would free all the session resources and all the session objects including the head (which is not the case currently).

Finally we don't have to change anything regarding `ssl->session_negotiate` but to copy in it some information (if any) about the session we eventually decide to resume.

Reply:
Agree with you. But I think we should change the type of  `resumption_session_list ` . I prefer bellow structure.

```
typedef struct 
{
   /* size and data should be store/load with mbedtls_ssl_session_{load,save} */
    void *resumption_session_data;
    size_t resumption_session_size;
    mbedtls_ssl_resumption_session_list *next;
} mbedtls_ssl_resumption_session_list;

mbedtls_ssl_handshake_params 
{
  ...
  mbedtls_ssl_resumption_session_list *resumption_session_list;
  ...
};
```

Reply:
Done. Please re-check that.